### PR TITLE
FOGL-7448 Update notification documentation

### PR DESCRIPTION
**Pull request by MarkRiddoch**
Signed-off-by: Mark Riddoch <mark@dianomic.com>

### CODE REVIEW

**Review by pintomax — APPROVED** [review]

**Review by ashish-jabble — CHANGES_REQUESTED** [review]

**Review by ashish-jabble — APPROVED** [review]

File: docs/index.rst [hunk 1]
```diff
@@ -27,14 +30,78 @@
...
+The following sections discuss each of the sources of data that may be
+used to cause evaluation of the notification rules.
+
+Reading Data
+------------
+
+The notification services observes data as it flows into the Fledge
+storage service buffer and processes that data against a set of rules that
+are configurable by the user to determine if an event has occurred. Events
+may be either when a condition that was previously not met being is,
```
Comment (praveen-garg) [inline docs/index.rst:66]: not met being is?

File: docs/index.rst [hunk 2]
```diff
@@ -27,14 +30,78 @@
...
+or a condition that was previously met becoming no longer true. The
+notification service can then send a notification when an event occurs
+or, in the case of a condition that is met, it can send notifications
+as long as that condition is met.
```
Comment (praveen-garg) [inline docs/index.rst:70]: can be written better

File: docs/index.rst [hunk 3]
```diff
@@ -27,14 +30,78 @@
...
+evaluate if a condition has occurred. Also the data that is observed
+by the notification is after any filtering rules have been applied in
+the south services but before any filtering that occurs in the north
+tasks. The mechanism used to allow the notification service to observe
+data is that the notifications register with the storage service to be
+given the values for particular assets as they arrive at the storage
+service. A notification may register for several assets and is free
+to buffer that data internally within the notification service. This
+registration does not impact how the data that is requested is treated
+in the rest of the system; it will still for example follow the normal
```
Comment (praveen-garg) [inline docs/index.rst:84]: it will still `for example` follow ?

File: docs/index.rst [hunk 4]
```diff
@@ -27,14 +30,78 @@
...
+processing rules to be sent onward to the north systems.
+
+Statistics Data
+---------------
+
+The notifications can register to obtain two variants of the
+statistics data, the raw statistic increments as they occur or the
```
Comment (praveen-garg) [inline docs/index.rst:91]: raw statistic increments or raw statistic incremented value?

File: docs/index.rst [hunk 5]
```diff
@@ -27,14 +30,78 @@
...
+
+Audit Data
+----------
+
+The notifications can register to obtain the |audit_trail| entries as they
+are raised by the system to the audit logger. These notifications receive
+data with the audit log name code as the data name and the data that is
+posted with the audit log entry as the data points of the data. There is a
+limited set of notification rule plugins that can be used with this data as
+it tends to be non-numeric and most plugins expect to sue numeric data.
```
Comment (praveen-garg) [inline docs/index.rst:104]: sue - > use*

File: docs/index.rst [hunk 6]
```diff
@@ -27,14 +30,78 @@
...
+configuration of the *stats collection* schedule. The default is to update
+the rate every 15 seconds and show the rates for 15 second intervals.
+
+Audit Data
+----------
+
+The notifications can register to obtain the |audit_trail| entries as they
+are raised by the system to the audit logger. These notifications receive
+data with the audit log name code as the data name and the data that is
+posted with the audit log entry as the data points of the data. There is a
```
Comment (praveen-garg) [inline docs/index.rst:102]: data points?

File: docs/index.rst [hunk 7]
```diff
@@ -294,3 +361,68 @@ A single page dialog appears that allows you to change any of the parameters of
...
+
+Statistics Rate Example
+-----------------------
+
+A simple example of the use of statistics for a notification is to
+monitor the flow of data within a pipeline. We can use the statistics
+that the north service updates for the data processed by the north
+service to determine if the service is functioning and sending data out
+of the Fledge instance. In this case the name of the statistic we want
+to monitor is the same as the name of the service.
```
Comment (praveen-garg) [inline docs/index.rst:373]: Can we use `Readings sent` for total sent as well? or only service name are allowed.

File: docs/index.rst [hunk 8]
```diff
@@ -294,3 +361,68 @@ A single page dialog appears that allows you to change any of the parameters of
...
+
+    +-----------+
+    | |ADHRule| |
+    +-----------+
+
+  - We select the *Data Source* as *Statistics History*, this will gives a a value every 15 seconds of the number of readings sent to ADH in the previous 15 seconds. We expect at least one value will be sent every 15 seconds.
+
+  - Select a condition of *<* and enter a value of *1*.
+   
+This will cause the notification to trigger if the value of the statistic
```
Comment (praveen-garg) [inline docs/index.rst:387]: value of the statistic rate

File: docs/index.rst [hunk 9]
```diff
@@ -294,3 +361,68 @@ A single page dialog appears that allows you to change any of the parameters of
...
+is less than 1. If we wanted to trigger on a low rather than 0 flow of
+data then we can obviously increase this value. Of course that is reliant
+on the user knowing what a reasonable value is. It might be better, if an
+alert is required when the flow drops of to use the 8Average* filter and
```
Comment (praveen-garg) [inline docs/index.rst:391]: 8Average* -> *Average*

File: docs/index.rst [hunk 10]
```diff
@@ -294,3 +361,68 @@ A single page dialog appears that allows you to change any of the parameters of
...
+
++--------------------+
+| |ADHRatePerMinute| |
++--------------------+
+
+We can configure any of the supported delivery plugins for our new
+notification dependent on how and who or what we would like to inform
+regarding this condition. It may be a simple message to an operator,
+raising a ticket in a ticketing system or issue some control message to
+cause fail over to a standby system.
```
Comment (praveen-garg) [inline docs/index.rst:408]: failover  (no space)

File: docs/index.rst [hunk 11]
```diff
@@ -294,3 +361,68 @@ A single page dialog appears that allows you to change any of the parameters of
...
+
+Audit Data Example
+------------------
+
+A simple example of a notification based on audit data would be to send a notification any time any service restarts. We do this by looking for particular audit codes, the audit codes that are supported by the system are defined in the |audit_trail| section.
+
+The audit code we will use in this case is SRVRG which is written whenever
```
Comment (praveen-garg) [inline docs/index.rst:415]: in this case -> in this example

File: docs/index.rst [hunk 12]
```diff
@@ -294,3 +361,68 @@ A single page dialog appears that allows you to change any of the parameters of
...
+a new service instance registers with the Fledge core.
+
+As with any notification we start by creating a naming the notification. Then we select the notification rule plugin. In this case we will use the *Data Availability* plugin. This rule triggers every time it receives data that matches the asset name or audit name it is given.
```
Comment (praveen-garg) [inline docs/index.rst:418]: creating a naming?

Comment (praveen-garg) [reply docs/index.rst:418]: audit name  -> audit code

File: docs/index.rst [hunk 13]
```diff
@@ -294,3 +361,68 @@ A single page dialog appears that allows you to change any of the parameters of
...
+
+We enter the required audit code in the *Audit Code* field. If we wanted to monitor several audit codes we could a comma separated list, however in this case one entry is sufficient.
```
Comment (praveen-garg) [inline docs/index.rst:420]: If we wanted to monitor several audit codes we could a comma separated list

If we want to monitor several audit codes we could enter a comma separated list e.g. SRVRG, SRVUN,

**Pull request merged**